src/Week7/HW07.hs

```diff
@@ -0,0 +1,197 @@
+
+partitionAt :: Ord a => Vector a -> Int -> (Vector a, a, Vector a)
+partitionAt vec i = (less, pivotVal, more)
+    where
+        pivotVal = vec ! i
+        (less,more) = V.partition (< pivotVal) vec
+
+
+-- Exercise 7 -----------------------------------------
+
+-- Quicksort
+quicksort :: Ord a => [a] -> [a]
+quicksort [] = []
+quicksort (x:xs) = quicksort [ y | y <- xs, y < x ]
+                   <> (x : quicksort [ y | y <- xs, y >= x ])
+
+qsort :: Ord a => Vector a -> Vector a
+qsort = V.fromList . quicksort . V.toList
+
+
+-- Exercise 8 -----------------------------------------
+
+qsortR :: Ord a => Vector a -> Rnd (Vector a)
+qsortR vec
+    | V.null vec = return V.empty
+    | otherwise = details >>= helperSort
+        where
+            details = getRandomR (0, V.length vec -1) >>= return . partitionAt vec
+            helperSort (l,m,h) = liftM2 (<>) (qsortR l) (Control.Monad.Random.liftM (cons m) (qsortR h))
+
+
+
+
+-- Exercise 9 -----------------------------------------
+
+-- Selection
+select :: Ord a => Int -> Vector a -> Rnd (Maybe a)
+select idx vec 
+    | V.null vec = return Nothing 
+    | otherwise = chosen >>= helper
+        where 
+            chosen = getRandomR (0, V.length vec -1) >>= return . partitionAt vec
+            helper (l,m,h)
+                | idx < lLen = select idx l
+                | idx > lLen = select (idx - lLen - 1) h
+                | otherwise = return $ Just m 
+                where 
+                    lLen = V.length l
+
+-- Exercise 10 ----------------------------------------
+
+allCards :: Deck
+allCards = [Card label suit | suit <- suits, label <- labels ]
+
+newDeck :: Rnd Deck
+newDeck =  shuffle allCards
+
+-- Exercise 11 ----------------------------------------
+
+nextCard :: Deck -> Maybe (Card, Deck)
+nextCard d 
+    | V.null d = Nothing 
+    | otherwise = Just (V.head d, V.tail d)
+
+-- Exercise 12 ----------------------------------------
+
+getCards :: Int -> Deck -> Maybe ([Card], Deck)
+getCards n d = helper n d []
+    where 
+        helper 0 d acc = return (acc,d)
+        helper n d acc = 
+                do
+                    (c,de) <- nextCard d
+                    helper (n-1) de (c : acc)
+
+-- Exercise 13 ----------------------------------------
+
+data State = State { money :: Int, deck :: Deck }
+
+repl :: State -> IO ()
+repl s@State{..} | money <= 0  = putStrLn "You ran out of money!"
+                 | V.null deck = deckEmpty
+                 | otherwise   = do
+  putStrLn $ "You have \ESC[32m$" ++ show money ++ "\ESC[0m"
+  putStrLn "Would you like to play (y/n)?"
+  cont <- getLine
+  if cont == "n"
+  then putStrLn $ "You left the casino with \ESC[32m$"
+           ++ show money ++ "\ESC[0m"
+  else play
+    where deckEmpty = putStrLn $ "The deck is empty. You got \ESC[32m$"
+                      ++ show money ++ "\ESC[0m"
+          play = do
+            putStrLn "How much do you want to bet?"
+            amt <- read <$> getLine
```

Comment:
There's also [readLn](https://hackage.haskell.org/package/base-4.15.0.0/docs/Prelude.html#v:readLn).
But similar to read it's partial (== crash at runtime if you enter non-number).
It's better to use [readMaybe](https://hackage.haskell.org/package/base-4.15.0.0/docs/Text-Read.html#v:readMaybe) and deal with the parse error explicitly.